### ISSUE TITLE
Order matters when configuring derived types with HashSet<int> in base type

### ISSUE THREAD

**Issue by mattsve** [dotnet/efcore#32745]
When updating to net8 and efcore8 we ran into an issue where we get runtime exceptions when trying to query the db. It works as intended in efcore7. I've managed to replicate the exception in a small example project. If the base type is configured after the derived types it works, but since we are using ApplyConfigurationsFromAssembly we cannot control the order.

The exception happens as soon as we query the database for a derived type.
```
System.ArgumentException
  HResult=0x80070057
  Message=Expression of type 'System.Collections.Generic.HashSet`1[System.Int32]' cannot be used for parameter of type 'System.Collections.Generic.IList`1[System.Int32]' of method 'System.Collections.Generic.IList`1[System.Int32] PopulateList[Int32](System.Collections.Generic.IList`1[System.Int32], System.Collections.Generic.IList`1[System.Int32])' (Parameter 'arg0')
  Source=System.Linq.Expressions
  StackTrace:
   at System.Dynamic.Utils.ExpressionUtils.ValidateOneArgument(MethodBase method, ExpressionType nodeKind, Expression arguments, ParameterInfo pi, String methodParamName, String argumentParamName, Int32 index)
   at System.Linq.Expressions.Expression.Call(MethodInfo method, Expression arg0, Expression arg1)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityMaterializerSource.<AddInitializeExpressions>g__CreateMemberAssignment|13_0(Expression parameter, MemberInfo memberInfo, IPropertyBase property, Expression value)  
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityMaterializerSource.AddInitializeExpressions(HashSet`1 properties, ParameterBindingInfo bindingInfo, Expression instanceVariable, List`1 blockExpressions)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityMaterializerSource.CreateMaterializeExpression(List`1 blockExpressions, ParameterExpression instanceVariable, Expression constructorExpression, HashSet`1 properties, ParameterBindingInfo bindingInfo)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityMaterializerSource.CreateMaterializeExpression(EntityMaterializerSourceParameters parameters, Expression materializationContextExpression)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.EntityMaterializerInjectingExpressionVisitor.CreateFullMaterializeExpression(ITypeBase concreteTypeBase, ValueTuple`4 materializeExpressionContext)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.EntityMaterializerInjectingExpressionVisitor.MaterializeEntity(StructuralTypeShaperExpression shaper, ParameterExpression materializationContextVariable, ParameterExpression concreteEntityTypeVariable, ParameterExpression instanceVariable, ParameterExpression entryVariable)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.EntityMaterializerInjectingExpressionVisitor.ProcessEntityShaper(StructuralTypeShaperExpression shaper)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.EntityMaterializerInjectingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.EntityMaterializerInjectingExpressionVisitor.Inject(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.InjectEntityMaterializers(Expression expression)
   at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryShapedQueryCompilingExpressionVisitor.ShaperExpressionProcessingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryShapedQueryCompilingExpressionVisitor.ShaperExpressionProcessingExpressionVisitor.ProcessShaper(Expression shaperExpression)
   at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryShapedQueryCompilingExpressionVisitor.VisitShapedQuery(ShapedQueryExpression shapedQueryExpression)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryShapedQueryCompilingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Program.<Main>$(String[] args) in D:\vt\experimental\eftest2\ConsoleApp1\Program.cs:line 6
```

```
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.ChangeTracking;
using Microsoft.EntityFrameworkCore.Storage.ValueConversion;

var dbContext = new MyDbContext();
dbContext.Set<Model1>().FirstOrDefault();

public class MyDbContext : DbContext
{
    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        // This order works
        // modelBuilder.Entity<Model1>();
        // modelBuilder.Entity<Model2>();
        // modelBuilder.Entity<BaseModel>().HasKey(m => m.Id);
        // modelBuilder.Entity<BaseModel>().Property(m => m.IntSet).HasConversion(new IntHashSetToStringValueConverter());
        // modelBuilder.Entity<BaseModel>().Property(m => m.IntSet).Metadata.SetValueComparer(new IntHashSetValueComparer());
        // modelBuilder.Entity<BaseModel>().HasDiscriminator<string>("Discriminator");

        // This order does not work
        modelBuilder.Entity<BaseModel>().HasKey(m => m.Id);
        modelBuilder.Entity<BaseModel>().Property(m => m.IntSet).HasConversion(new IntHashSetToStringValueConverter());
        modelBuilder.Entity<BaseModel>().Property(m => m.IntSet).Metadata.SetValueComparer(new IntHashSetValueComparer());
        modelBuilder.Entity<BaseModel>().HasDiscriminator<string>("Discriminator");
        modelBuilder.Entity<Model1>();
        modelBuilder.Entity<Model2>();
    }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) =>
        optionsBuilder.UseInMemoryDatabase("dbName");
}

public abstract class BaseModel
{
    public int Id { get; set; }
    public HashSet<int> IntSet { get; set; } = new();
}

public class Model1 : BaseModel;

public class Model2 : BaseModel;

public class IntHashSetToStringValueConverter() 
    : ValueConverter<HashSet<int>, string>(x => ToString(x), x => ToHashSet(x))
{
    public static string ToString(ICollection<int> collection)
    {
        var strings = collection.Select(i => i.ToString());
        return string.Join(",", strings);
    }

    public static HashSet<int> ToHashSet(string @string)
    {
        var parts = @string.Split(',').Where(p => !string.IsNullOrWhiteSpace(p));
        var values = parts.Select(int.Parse);
        return values.ToHashSet();
    }
}

public class IntHashSetValueComparer()
    : ValueComparer<HashSet<int>>((c1, c2) => c1!.SequenceEqual(c2!),
        c => c.Aggregate(0, (a, v) => HashCode.Combine(a, v.GetHashCode())),
        c => c.ToHashSet());
```

**Comment by ajcvickers** [comment]
Note for triage: this is working on the latest daily build, but still broken in 8.0.1. I have not yet been able to determine if it is fixed or not in 8.0.2.

**Comment by ajcvickers** [comment]
Same root cause as #32411, which is fixed in 8.02.

**Comment by kawazoe** [comment]
I'd like to point out that this issue also happens if you're using the Npgsql client with EnableDynamicJson. All my jsonb fields declared as ICollection or ISet stopped working with a similar error.
